Scada interface

WenzDaniel opened the pull request:
**What is the problem / what does the code in this PR do**
With this PR we implements a simple python-scada interface like in hax.

**Can you briefly describe how it works?**
We access the scada database via the two web APIs GetSCLastValue and GetSCData. SCADA stores only values which changes. Hence when GetSCData is used to query a time range in which for the given parameter no value had been stored, nothing is returned by the API. To circumvent this problem the data is read as follows:

1. An empty pandas DataFrame is initialized with a 1 sec spacing (sampling frequency of SCADA).
2. GetSCLastValue is called to check when the last value of the specified parameter was recorded, given the time range queried by the user. (star)
3. GetSCData is used to get all data which had been recorded within the given time interval. This data is filled into the DataFrame.
4. All empty values in the DataFrame are filled via forward filling.

(star) this point requires a change in the GetSCLastValue  and is currently not yet implemented.

In order to have a convenient way to access the time ranges via a run_id I had to do some minor modifications in strax which are detailed here https://github.com/AxFoundation/strax/pull/318.  

**Can you give a minimal working example (or illustrate with a figure)?**
A MWE can be found in this [notebook](https://github.com/XENONnT/analysiscode/blob/master/StraxTests/ScadaInterface.ipynb)

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,246 @@
+        # by the scada api is always the same...
+        temp_df = pd.read_json(values.text)
+        df.loc[temp_df['timestampseconds'], parameter_key] = temp_df.loc[:, 'value'].values
+    except:
```
Too broad an exception, can we be a little more specific?

WenzDaniel replied:
Okay done

jorana replied:
@WenzDaniel I don't think you did not?

WenzDaniel replied:
I do not know what you mean I nailed it down to ValueError?

jorana replied:
Perhaps I looked at the wrong commit history, there are quite a few ;) but looks good now!

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,246 @@
+
+
+@export
+def convert_time_zone(df, tz):
```
Why not make 'strax' the default?

jorana replied:
So your argument is to better have people aware than making the choice for them, I buy that.

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,246 @@
+
+
+@export
+def get_scada_values(parameters,
```
Perhaps you thought about it but would a Scada class not make somewhat sense?

WenzDaniel replied:
Yes I thought about this, but I think there is not much of a benefit here. It would allow us to  move line 10 - 18 into the init. I am fine with both.

jorana replied:
Perhaps as the code gets more complex, certain things are nice to set in a class (like the tz on the init). Perhaps it's easier to already set it up now such that the code does not change all the time. Perhaps the advantage is limited now but may end up significant later.

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,246 @@
+        if ind:
+            m = np.all(df.loc[:, 'time'] == temp_df.loc[:, 'time'])
+            mes = ('This is odd somehow the time stamps for the query of'
+                   f' {p} does not match the other time stamps.')
```
This can happen right if the loop takes long and you specify a time in the future?

jorana reviewed (commented):
Thanks Daniel. Generally looks good, have not tested it but looks relatively straightforward and very useful!

WenzDaniel commented:
Good, catch with the "future" values. It does not fail, but I think in case the user asks for some future values the minimal thing we should do is to print a warning, but I think it would be even better to return nans and the user shall decide how to project values into the future.

WenzDaniel commented:
Okay implemented all comments and updated the example notebook.

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,275 @@
+        except ValueError:
+            raise ValueError('Cannot load SCADA information, from xenon'
+                             ' secrets. SCADAInterface cannot be used.')
+        self.st = context
```
self.context seems cleaner

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,275 @@
+        if np.all((run_id, self.st)):
+            # User specified a valid context and run_id, so get the start
+            # and end time for our query:
+            if isinstance(run_id, (list, tuple)):
```
Why not string?

WenzDaniel replied:
Because, this is only needed if the user specifies a list/tuple of run_ids.

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,275 @@
+        :param start: int representing the start time of the interval in ns
+            unix time.
+        :param end: same as start but as end.
+        :param run_id: Id of the run. Can also be specified as a list or
```
tuble -> tuple 

also not restricted to 2 right?

jorana commented on straxen/scada.py:
```diff
@@ -0,0 +1,275 @@
+
+        now = np.datetime64('now')
+        if (end // 10**9) > now.astype(np.int64):
+            mes = ('You are asking for an endtime which is in the future,'
```
Nice

jorana reviewed (approved):
Awesome code, well done Daniel! Feel free to merge if you want to merge this PR if you think you addressed the comments sufficiently.

WenzDaniel commented:
Okay fixed all comments, also added the possibility to interpolate between data points and to "down sample" the data (picking every nth point). I also updated the example accordingly.
